src/main/java/org/javarosa/core/model/instance/CsvExternalInstance.java

```diff
@@ -1,38 +1,52 @@
 package org.javarosa.core.model.instance;
 
 import java.io.BufferedReader;
+import java.io.File;
 import java.io.FileReader;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
 import org.javarosa.core.model.data.UncastData;
 
 public class CsvExternalInstance {
     public static TreeElement parse(String instanceId, String path) throws IOException {
-        TreeElement root = new TreeElement("root", 0);
+        final TreeElement root = new TreeElement("root", 0);
         root.setInstanceName(instanceId);
 
-        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
-            String csvLine = br.readLine();
+        final CSVFormat csvFormat = CSVFormat.DEFAULT
+            .withDelimiter(getDelimiter(path))
+            .withFirstRecordAsHeader();
+        final CSVParser csvParser = CSVParser.parse(new File(path), StandardCharsets.UTF_8, csvFormat);
+        final String[] fieldNames = csvParser.getHeaderMap().keySet().toArray(new String[0]);
+        int multiplicity = 0;
 
-            if (csvLine != null) {
-                String[] fieldNames = csvLine.split(",");
-                int multiplicity = 0;
+        for (CSVRecord csvRecord : csvParser.getRecords()) {
+            TreeElement item = new TreeElement("item", multiplicity);
 
-                while ((csvLine = br.readLine()) != null) {
-                    TreeElement item = new TreeElement("item", multiplicity);
-                    String[] data = csvLine.split(",");
-                    for (int i = 0; i < fieldNames.length; ++i) {
-                        TreeElement field = new TreeElement(fieldNames[i], 0);
-                        field.setValue(new UncastData(i < data.length ? data[i] : ""));
-
-                        item.addChild(field);
-                    }
-
-                    root.addChild(item);
-                    multiplicity++;
-                }
+            for (int i = 0; i < fieldNames.length; ++i) {
+                TreeElement field = new TreeElement(fieldNames[i], 0);
+                field.setValue(new UncastData(i < csvRecord.size() ? csvRecord.get(i) : ""));
+                item.addChild(field);
             }
+
+            root.addChild(item);
+            multiplicity++;
         }
 
         return root;
     }
+
+    private static char getDelimiter(String path) throws IOException {
+        char delimiter = ',';
+        try (BufferedReader reader = new BufferedReader(new FileReader(path))) {
+            String header = reader.readLine();
+
+            if (header.contains(";")) {
```

Comment:
I guess this means a CSV like `hello; world, blah` would have columns `hello` and `world, blah` rather than `hello; world` and `blah` (which I might have meant). I think that's ok, as parsing these things with more than one delimiter and always getting it right is impossible, but just thinking out loud.

Reply:
I meant to call this out, sorry! It feels like having punctuation in column headers would be very unusual. I can kind of imagine it with commas maybe? Something cruel like a column header of `"First, Last Name"`? But I really can't imagine using semicolons in a column header so that's why it felt safest to explicitly test for semicolon. If we really wanted to be fancy we could try to account for quotes in headers but that really felt like overkill. I'd like to go with the limitation you pointed out and adjust if someone has a legitimate usecase for a semicolon in the header (please, no!).

Reply:
Which reminds me of something else I meant to mention -- I briefly considered also supporting tabs but again, I think we should wait until someone explicitly asks for it. The C in CSV is comma.

Reply:
> The C in CSV is comma.

This.